openeogeotrellis/integrations/identity.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import json
+import logging
+import os
+from dataclasses import dataclass
+from datetime import timedelta
+from pathlib import Path
+from typing import Literal, Optional
+
+import jwt
+
+from openeogeotrellis.config import get_backend_config
+from openeogeotrellis.utils import get_file_reload_register_func_if_changed, utcnow
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class IDPDetails:
+    """Identity Provider Details to create identities"""
+    issuer: str
+    """URI used to symbolize the issuer of the issued tokens"""
+    private_key: str
+    """The private key in PEM format used to sign the tokens"""
+    public_key: str
+    """The public key in PEM format which can be used to verify the tokens"""
+
+    @classmethod
+    def from_file(cls, file_path: Path) -> "IDPDetails":
+        with open(file_path) as fh:
+            file_dict = json.load(fh)
+        private_key = file_dict["private_key"].strip()
+        cls.assert_valid_pem(private_key, "private")
+        public_key = file_dict["public_key"].strip()
+        cls.assert_valid_pem(public_key, "public")
+        return cls(
+            file_dict["issuer"],
+            private_key,
+            public_key
+        )
+
+    @staticmethod
+    def assert_valid_pem(key: str, key_type: Literal["private", "public"]):
+        """Check if a string follows the expected structure of a PEM formatted key"""
+        template = '-----{location} RSA {key_type} KEY-----'
+        pem_header = template.format(location="BEGIN", key_type=key_type.upper())
+        pem_footer = template.format(location="END", key_type=key_type.upper())
+        assert key.startswith(pem_header), f"Key did not start with {pem_header}"
+        assert key.endswith(pem_footer), f"Key did not end with {pem_footer}"
+
+
+class IDPTokenIssuer:
+    """
+    An issuer of signed tokens. This is a singleton so to avoid errors having duplicate instances you can call the
+    class method `instance() to get the singleton object`.
+    """
+    _TOKEN_EXPIRES_IN_SECONDS = 48 * 3600
+    _SINGLETON: Optional[IDPTokenIssuer] = None
+
+    def __init__(self):
+        if self.__class__._SINGLETON is not None:
+            raise RuntimeError("IDPTokenIssuer is meant as a singleton")
+        self._IDP_DETAILS: Optional[IDPDetails] = None
+        self.__class__._SINGLETON = self
```

Comment:
I prefer it to be a singleton for the time being. The singleton check is not a one-way door as it can be easily reverted. Having 2 issuers on the other end creates a bunch of problems (especially with well-known paths once it is preferred to public key material.

Reply:
I'm still confused. The main defining feature I see here is to reload some data from a file if you observe it has been touched since last read. But I don't understand how that requires a singleton.

And related: why would we want to hardcode here that there can only be one issuer in the application? For example: in openEO auth we support multiple OIDC providers at the same time

Reply:
The reloading of the config does not require it to be a singleton as long as there would be a watcher per instance. The reason why it is a singleton is because there should not be a need for multiple IDPTokenIssuer's and if there is it wouldn't be hard to allow it.

Yes we support different OIDC providers for auth such that users don´t have to register themselves with us but we should not create different identity providers for tokens that we issue ourselves. If there are tokens required for different purposes that would still be possible within this singleton by adding limitations in a standard way (e.g. allowing multiple types of tokens and adding an audience for certain type of tokens that should be limited to specific audiences) but the issuer should always be a reference to the OpenEO instance.

So given that that is logically 1 instance for a given runtime a singleton is preferred.